src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -0,0 +1,487 @@
+            top_pad: The number of pixels to pad the top of the block.
+            bottom_pad: The number of pixels to pad the bottom of the block.
+
+        Returns:
+            TextBlock: text block with the text set.
+            str: the language of the text or None if the OCR processor doesn't support language detection.
+        """
+        # TODO: THis won't work currently because the image isn't part of the class.
+        # Pad to improve OCR accuracy as it's fairly tight.
+
+        padded_block = block.pad(
+            left=left_pad, right=right_pad, top=top_pad, bottom=bottom_pad
+        )
+
+        segment_image = padded_block.crop_image(image)
+
+        # Perform OCR
+        if isinstance(self.ocr_agent, TesseractAgent):
+            language = None
+            text = self.ocr_agent.detect(segment_image, return_only_text=True)
+        elif isinstance(self.ocr_agent, GCVAgent):
+            gcv_response = self.ocr_agent.detect(segment_image, return_response=True)
+            text = gcv_response.full_text_annotation.text
+
+            # We assume one language per text block here which seems reasonable, but may not always be true.
+            try:
+                language = (
+                    gcv_response.full_text_annotation.pages[0]
+                    .property.detected_languages[0]
+                    .language_code
+                )
+            except IndexError:
+                # No language was found in the GCV response
+                language = None
+        else:
+            raise ValueError(
+                "The OCR agent must be either a TesseractAgent or a GCVAgent."
+            )
+
+        # Save OCR result
+        block_with_text = padded_block.set(text=text)  # type: ignore
+
+        return block_with_text, language  # type: ignore
+
+    @staticmethod
+    def _is_block_valid(block: TextBlock) -> bool:
+        """Check if a block is valid."""
+        if block.text is None:
+            return False
+        if len(block.text) == 0:
+            return False
+        # Heuristic to get rid of blocks with no text or text that is too short.
+        if block.type == "Inferred from gaps":
+            if len(block.text.split(" ")) < 3:
+                return False
+        return True
+
+    def process_layout(self) -> Tuple[List[PDFTextBlock], List[TextBlock]]:
+        """Get text for blocks in the layout and return a `Page` with text, language id per text block
+
+        :return: list of text blocks with text, language and text block IDs set + a list of text blocks in
+        layoutparser's format (useful for visual debugging).
+        """
+        text_blocks, text_layout = [], []
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            for block_idx, block in enumerate(self.layout):
+                text_block_id = f"p_{self.page_number}_b_{block_idx}"
+                # Skip blocks that already have text (likely because we used google's structure detection)
```

Comment:
I moved some files around here so this is all green, but the only change I made is here (skipping API calls for any block that already has text from an API call on a whole page).